### ISSUE TITLE
Security Disclosure from HP Fortify

### ISSUE THREAD

**Issue by shadowhand** [kohana/kohana#74]
The following report just landed in my inbox. It is a security analysis done as part of the HP Fortify project. It outlines a number of (potentially) serious flaws in Kohana 3.3.3.1. I scanned the report and some of the concerns raised are in modules that would not (hopefully) be used in production, including some unit tests.

https://www.dropbox.com/s/6cw6gmq5l6vfqnj/Kohana%203331.pdf?dl=0

Here is the text of the email:

---

Dear Project Manager,

History of Fortify Open Source Review

Since 2001, HP Fortify, the leader in application security testing, has dedicated our market leading Source Code Analyzer (SCA) solution to the advancement and security of Open Source security projects.  Today, as HP Fortify continues its journey, the HP Fortify Open Review team is providing open source project owners a no-cost assessment.  Contributors are provided every opportunity to not only provide the community with great software, but secure software.

Open Source Project / Application Name and release version:  Kohana 3.3.3.1

The HP Fortify Open Review team has assessed Kohana 3.3.3.1 for possible security vulnerabilities and the results of your assessment is attached.  It is HP’s policy to make all results public on our Fortify on Demand website within 60 days from the date of this notification. Based on the findings, we would encourage your team to remediate any of the security vulnerabilities in this report or challenge any finding as a “false positive” by contacting our team with an explanation of why you believe the finding to be false.

To contact to a member of our team, please email us at Fortify-Open-Review@hp.com.  

Additional information about Hewlett-Packard’s Fortify Open Review program is available here: 

https://hpfod.com/open-source-review-project

What is Fortify on Demand?

HP Fortify on Demand is a managed application security testing service that makes it simple to  initiate security tests on a few applications or launch a comprehensive security program without  upfront investment of technology  and resources. Combining advanced dynamic and static  testing technologies (HP Fortify) with HP’s experience in evaluating software security, Fortify on  Demand brings professional-level software security expertise to organizations of any size.

Regards,

The Fortify on Demand Team
© 2015 Hewlett-Packard Development Company, L.P.

**Comment by pocesar** [comment]
good to know that the non-dev part of Kohana is still solid :+1: 
I've never been too fond of having extract() inside the database module drivers though

**Comment by enov** [comment]
Thanks @shadowhand, and thank you HP for the security review.

Skimming through this, it seems to me that this is probably an automated (or half-automated) code analysis. Personally, I did not find anything to be panicked about, but needless to say, we take such reviews very seriously.

I have created a label "HP Fortify" and attached it to this issue. We'll have to create and apply it across the repositories on the issues related to this. This is for future reference, as well as to make it easy for HP personnel to track this, in case there will be follow-up reviews, if any.

**Comment by JackEllis** [comment]
I agree with the password hardcoding part somewhat but I don't see an issue with having a 'default' database.php config file and then ignoring it / overriding it during deployment. There are many set-ups that are doable and I feel like this is the responsibility of the developer implementing things and not the Kohana team?

**Comment by enov** [comment]
@JackEllis I feel like so.

On the other side, now that I am reviewing things thoroughly, I think there is a serious issue in trusting `$_SERVER['HTTP_HOST']`, even though it is not easily exploitable.

**Comment by acoulton** [comment]
@JackEllis both to encourage good practice, and because all the groups of the default config file get merged into the end-user's application, I think it would be better if we shipped the database config file as an example with all sections commented out, the same way we do for the cache module. That would be a breaking change, obviously.

**Comment by farazmunj** [comment]
I know this issue is closed, but I have a question. I am adding www.domainname and domainname in site config but some time server receives requests with where HTTP_HOST value as www.domainname:80 and it get rejected by Kohana.  
So my question is should I add this value in URL config and trust this request or keep rejecting them. Most requests are from web crawlers.

**Comment by acoulton** [comment]
@farazilu I'd say it depends whether you want to serve those requests or not. It's valid to include the port number in the host header, but all the mainstream browsers and request libraries strip the :80 default port. It probably suggests a crawler using a nonstandard library - up to you whether you care about rejecting them. If you want to serve those clients add it to your regex, otherwise leave it.
